onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -193,13 +186,23 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
       base_offset = compute_base_offset(process_dims, input_shape_pitches, axis);
 
       // process 1 chunk of 'inner dimension' length
-      for (int64_t i = 0; i < inner_dim_size; ++i) {
-        // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
-        // optimizer will remove the redundant if/else block based on 'is_string' template parameter
-        if (is_string) {
-          output_data[++output_counter] = input_data[base_offset + indices_data[++indices_counter]];
-        } else {
-          memcpy(output_data, input_data + (base_offset + indices_data[++indices_counter]) * element_size, element_size);
+      if (is_string) {
+        for (int64_t i = 0; i < inner_dim_size; ++i) {
+          // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
+          // optimizer will remove the redundant if/else block based on 'is_string' template parameter
+          output_data[++output_counter] =
+              input_data[base_offset +
+                         GetNegativeIndexAdjustedValue<Tin>(indices_data, ++indices_counter, axis, input_shape)];
+        }
+      } else {
+        for (int64_t i = 0; i < inner_dim_size; ++i) {
```

Comment:
For 500k - 1M indices, we can potentially get more gains by parallelizing this loop ?

Reply:
Yeah, parallelizing this loop is pending. It'll take a little bit more work. I wanted to checkin first what I had working.